### PR TITLE
Add the METALC language

### PR DESCRIPTION
**Pull request by rwy0717**
Signed-off-by: Robert Young <rwy0717@gmail.com>

### CODE REVIEW

**Comment by rwy0717** [comment]
I think this is going to work as a very primitive first step towards compiling metal-c with cmake.

Note the following issues:
1. I have not provided a command to link a purely metal-c library.
2. Includes and compile options are ignored when compiling metal-c sources.

**Comment by fjeremic** [comment]
Will these restrictions affect us for what we're trying to do in OMR/OpenJ9?

**Comment by rwy0717** [comment]
Today, there's no need for it, but I can already see in https://github.com/eclipse/omr/pull/3923, that we will need a way to specify global includes for HLASM assembly.

The problem with using a custom language for METAL-C is, sources only have a single set of standard properties for configuring it's defines/includes/flags. However, the METAL-C pipeline has two stages, c->asm (`xlc`), and asm->object (`as`), and will need two distinct sets of options.

In my opinion, a target's standard properties should only apply to the `xlc` command, and we can provide some alternative properties (maybe prefixed with hlasm_) for influencing the `as` command.

So for now, I'm going to do two things:
1. Add a global cmake variable for adding flags to the `as` command
2. Ensure standard target defines/includes/flags apply to the xlc command.

(1) should be enough for eclipse/omr#3923 to work, and we can add finer-grained options if needed.

**Comment by fjeremic** [comment]
>     1. Add a global cmake variable for adding flags to the `as` command

Just clarifying, you mean the `as` command that gets executed as part of the METAL-C pipeline, correct?

**Comment by rwy0717** [comment]
>Just clarifying, you mean the as command that gets executed as part of the METAL-C pipeline, correct?

Yep. I just pushed a version that supports defines/includes/flags on the xlc command line, and users can set the variable `CMAKE_HLASM_FLAGS` to add options to the `as` command line.

File: Modules/CMakeMETALCInformation.cmake [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+set(CMAKE_METALC_SOURCE_FILE_EXTENSIONS mc)
+set(CMAKE_METALC_OUTPUT_EXTENSION .o)
+set(CMAKE_INCLUDE_FLAG_METALC "-I")
+set(CMAKE_METALC_DEFINE_FLAG "-D")
+
+set(CMAKE_METALC_COMPILE_OBJECT
+  "<CMAKE_COMMAND> -E copy <SOURCE> <OBJECT>.c"
+  "<CMAKE_METALC_COMPILER> -qmetal -qlongname -S <DEFINES> <INCLUDES> <FLAGS> -o <OBJECT>.s <OBJECT>.c"
+  "<CMAKE_HLASM_AS> -mgoff -I CBC.SCCNSAM <CMAKE_HLASM_FLAGS> -o <OBJECT> <OBJECT>.s"
```
Comment (fjeremic) [inline Modules/CMakeMETALCInformation.cmake]: What is the `-I CBC.SCCNSAM` for?

Comment (fjeremic) [reply Modules/CMakeMETALCInformation.cmake]: Oh I see. It looks to be needed for the prologue/epilog according to [1]. This should be fine.

[1] https://www.ibm.com/developerworks/community/forums/html/topic?id=85097a94-0889-4a80-80ed-4c052164dc22&ps=50

**Comment by fjeremic** [comment]
@rwy0717 looks good! Thanks for this. Anything else that needs to be done for this one? Once I get a confirmation I'll build with your change and just verify OMR builds are working as intended and we can merge and I'll update the new CMake versions on the build farm.

**Comment by fjeremic** [comment]
This seems to be replaced by the work done in [1] for now. See [2] for explanation as to why going to custom command route is more appropriate in the OMR case. I'm going to close this PR now. Please reopen if there is more work/discussion to be done here. Thanks for all the work in this area!

[1] https://github.com/eclipse/omr/pull/4117
[2] https://github.com/eclipse/omr/pull/4117#issuecomment-514673319

**Pull request closed without merging**